base/sparse/sparsevector.jl

```diff
@@ -0,0 +1,1441 @@
+### common.jl
+
+import Base: Func, AddFun, MulFun, MaxFun, MinFun, SubFun
+
+immutable RealFun <: Func{1} end
+call(::RealFun, x) = real(x)
+
+immutable ImagFun <: Func{1} end
+call(::ImagFun, x) = imag(x)
+
+immutable ComplexFun <: Func{2} end
+call(::ComplexFun, x::Real, y::Real) = complex(x, y)
+
+immutable DotFun <: Func{2} end
+call(::DotFun, x::Number, y::Number) = conj(x) * y
+
+typealias UnaryOp Union{Function, Func{1}}
+typealias BinaryOp Union{Function, Func{2}}
+
+### sparsevec.jl
+
+### Types
+
+immutable SparseVector{Tv,Ti<:Integer} <: AbstractSparseVector{Tv,Ti}
+    n::Int              # the number of elements
+    nzind::Vector{Ti}   # the indices of nonzeros
+    nzval::Vector{Tv}   # the values of nonzeros
+
+    function SparseVector(n::Integer, nzind::Vector{Ti}, nzval::Vector{Tv})
+        n >= 0 || throw(ArgumentError("The number of elements must be non-negative."))
+        length(nzind) == length(nzval) ||
+            throw(ArgumentError("index and value vectors must be the same length"))
+        new(convert(Int, n), nzind, nzval)
+    end
+end
+
+SparseVector{Tv,Ti}(n::Integer, nzind::Vector{Ti}, nzval::Vector{Tv}) =
+    SparseVector{Tv,Ti}(n, nzind, nzval)
+
+### Basic properties
+
+length(x::SparseVector) = x.n
+size(x::SparseVector) = (x.n,)
+nnz(x::SparseVector) = length(x.nzval)
+countnz(x::SparseVector) = countnz(x.nzval)
+nonzeros(x::SparseVector) = x.nzval
+nonzeroinds(x::SparseVector) = x.nzind
+
+similar{T}(x::SparseVector, ::Type{T}, D::Dims) = spzeros(T, D...)
+
+### Construct empty sparse vector
+
+spzeros(len::Integer) = spzeros(Float64, len)
+spzeros{T}(::Type{T}, len::Integer) = SparseVector(len, Int[], T[])
+
+### Construction from lists of indices and values
+
+function _sparsevector!{Ti<:Integer}(I::Vector{Ti}, V::Vector, len::Integer)
+    # pre-condition: no duplicate indexes in I
+    if !isempty(I)
+        p = sortperm(I)
+        permute!(I, p)
+        permute!(V, p)
+    end
+    SparseVector(len, I, V)
+end
+
+function _sparsevector!{Tv,Ti<:Integer}(I::Vector{Ti}, V::Vector{Tv}, len::Integer, combine::BinaryOp)
+    if !isempty(I)
+        p = sortperm(I)
+        permute!(I, p)
+        permute!(V, p)
+        m = length(I)
+
+        # advances to the first non-zero element
+        r = 1     # index of last processed entry
+        while r <= m
+            if V[r] == zero(Tv)
+                r += 1
+            else
+                break
+            end
+        end
+        r > m && SparseVector(len, Ti[], Tv[])
```

Comment:
@mbauman Did I miss something, it seems that that statement must be incorrect?
It creates a SparseVector and just throws it away.